File: src/CircleAI.Integration.Geo/OpenMeteoWeatherProvider.cs
// OpenMeteoWeatherProvider.cs
//
// (Phase B4) Open-Meteo free, no-API-key weather provider. Returns
// current conditions + hourly forecast in plain JSON.

using System;
using System.Collections.Generic;
using System.Globalization;
using System.Net.Http;
using System.Text.Json;
using System.Threading;
using System.Threading.Tasks;
using CircleAI.Integration;

namespace CircleAI.Integration.Geo;

public sealed class OpenMeteoWeatherProvider : IWeatherProvider
{
    private readonly HttpClient _http;

    public OpenMeteoWeatherProvider() : this(new HttpClient()) { }
    public OpenMeteoWeatherProvider(HttpClient http) => _http = http ?? throw new ArgumentNullException(nameof(http));

    public string ProviderId => "open-meteo";

    public async ValueTask<WeatherSample> CurrentAsync(double lat, double lon, CancellationToken ct = default)
    {
        var url = $"https://api.open-meteo.com/v1/forecast?latitude={lat.ToString(CultureInfo.InvariantCulture)}&longitude={lon.ToString(CultureInfo.InvariantCulture)}"
                + "&current=temperature_2m,apparent_temperature,precipitation,wind_speed_10m,cloud_cover,weather_code";
        using var resp = await _http.GetAsync(url, ct).ConfigureAwait(false);
        resp.EnsureSuccessStatusCode();
        using var doc = await JsonDocument.ParseAsync(
            await resp.Content.ReadAsStreamAsync(ct).ConfigureAwait(false),
            cancellationToken: ct).ConfigureAwait(false);
        var cur = doc.RootElement.GetProperty("current");
        var ts  = cur.GetProperty("time").GetString();
        return new WeatherSample(
            AtUtc:      DateTimeOffset.Parse(ts ?? DateTime.UtcNow.ToString("O"), CultureInfo.InvariantCulture, DateTimeStyles.AssumeUniversal).ToUniversalTime(),
            TempC:      cur.GetProperty("temperature_2m").GetDouble(),
            FeelsLikeC: cur.GetProperty("apparent_temperature").GetDouble(),
            PrecipMm:   cur.GetProperty("precipitation").GetDouble(),
            WindKph:    cur.GetProperty("wind_speed_10m").GetDouble() * 3.6, // m/s → km/h
            CloudPct:   cur.GetProperty("cloud_cover").GetInt32(),
            Condition:  WmoDecode(cur.GetProperty("weather_code").GetInt32()));
    }

    public async ValueTask<IReadOnlyList<WeatherSample>> HourlyAsync(double lat, double lon, int hours, CancellationToken ct = default)
    {
        if (hours <= 0 || hours > 168) throw new ArgumentOutOfRangeException(nameof(hours));
        var url = $"https://api.open-meteo.com/v1/forecast?latitude={lat.ToString(CultureInfo.InvariantCulture)}&longitude={lon.ToString(CultureInfo.InvariantCulture)}"
                + "&hourly=temperature_2m,apparent_temperature,precipitation,wind_speed_10m,cloud_cover,weather_code"
                + $"&forecast_hours={hours}";
        using var resp = await _http.GetAsync(url, ct).ConfigureAwait(false);
        resp.EnsureSuccessStatusCode();
        using var doc = await JsonDocument.ParseAsync(
            await resp.Content.ReadAsStreamAsync(ct).ConfigureAwait(false),
            cancellationToken: ct).ConfigureAwait(false);
        var h    = doc.RootElement.GetProperty("hourly");
        var time = h.GetProperty("time");
        var temp = h.GetProperty("temperature_2m");
        var feel = h.GetProperty("apparent_temperature");
        var prec = h.GetProperty("precipitation");
        var wind = h.GetProperty("wind_speed_10m");
        var cld  = h.GetProperty("cloud_cover");
        var code = h.GetProperty("weather_code");
        var n    = Math.Min(time.GetArrayLength(), hours);
        var result = new List<WeatherSample>(n);
        for (var i = 0; i < n; i++)
        {
            result.Add(new WeatherSample(
                AtUtc:      DateTimeOffset.Parse(time[i].GetString() ?? "", CultureInfo.InvariantCulture, DateTimeStyles.AssumeUniversal).ToUniversalTime(),
                TempC:      temp[i].GetDouble(),
                FeelsLikeC: feel[i].GetDouble(),
                PrecipMm:   prec[i].GetDouble(),
                WindKph:    wind[i].GetDouble() * 3.6,
                CloudPct:   cld[i].GetInt32(),
                Condition:  WmoDecode(code[i].GetInt32())));
        }
        return result;
    }

    /// <summary>(Phase B4) Decode WMO weather code (Open-Meteo standard).</summary>
    private static string WmoDecode(int code) => code switch
    {
        0          => "clear sky",
        1 or 2 or 3 => "partly cloudy",
        45 or 48   => "fog",
        51 or 53 or 55 => "drizzle",
        56 or 57   => "freezing drizzle",
        61 or 63 or 65 => "rain",
        66 or 67   => "freezing rain",
        71 or 73 or 75 => "snow",
        77         => "snow grains",
        80 or 81 or 82 => "rain showers",
        85 or 86   => "snow showers",
        95         => "thunderstorm",
        96 or 99   => "thunderstorm with hail",
        _          => "unknown",
    };
}

File: python/src/circle_ai/integration_geo/open_meteo_weather_provider.py
# open_meteo_weather_provider.py
#
# Port of CircleAI.Integration.Geo/OpenMeteoWeatherProvider.cs (C# — the EXACT
# spec).
#
# (Phase B4) Open-Meteo free, no-API-key weather provider. Returns current
# conditions + hourly forecast in plain JSON.
#
# The C# takes an injected ``HttpClient``; the Python port takes an injected
# :class:`IHttpFetcher` and parses the identical JSON shape. Wind is converted
# m/s -> km/h (``* 3.6``) exactly as the C#.

from __future__ import annotations

from typing import List

from circle_ai.integration._util import parse_utc
from circle_ai.integration.contracts import IWeatherProvider, WeatherSample
from circle_ai.integration.http import HttpRequest, IHttpFetcher


def _invariant(x: float) -> str:
    """C# ``double.ToString(CultureInfo.InvariantCulture)`` — '.' decimal, no
    thousands separators, no trailing ``.0`` for integers.
    """
    if x == int(x):
        return str(int(x))
    return repr(x)


def _wmo_decode(code: int) -> str:
    """(Phase B4) Decode WMO weather code (Open-Meteo standard). Mirrors the C#
    ``WmoDecode`` switch expression exactly.
    """
    if code == 0:
        return "clear sky"
    if code in (1, 2, 3):
        return "partly cloudy"
    if code in (45, 48):
        return "fog"
    if code in (51, 53, 55):
        return "drizzle"
    if code in (56, 57):
        return "freezing drizzle"
    if code in (61, 63, 65):
        return "rain"
    if code in (66, 67):
        return "freezing rain"
    if code in (71, 73, 75):
        return "snow"
    if code == 77:
        return "snow grains"
    if code in (80, 81, 82):
        return "rain showers"
    if code in (85, 86):
        return "snow showers"
    if code == 95:
        return "thunderstorm"
    if code in (96, 99):
        return "thunderstorm with hail"
    return "unknown"


class OpenMeteoWeatherProvider(IWeatherProvider):
    """Port of ``CircleAI.Integration.Geo.OpenMeteoWeatherProvider``."""

    def __init__(self, http: IHttpFetcher) -> None:
        if http is None:
            raise ValueError("http must not be None")
        self._http = http

    @property
    def provider_id(self) -> str:
        return "open-meteo"

    async def current_async(self, lat: float, lon: float) -> WeatherSample:
        url = (
            "https://api.open-meteo.com/v1/forecast"
            f"?latitude={_invariant(lat)}&longitude={_invariant(lon)}"
            "&current=temperature_2m,apparent_temperature,precipitation,"
            "wind_speed_10m,cloud_cover,weather_code"
        )
        resp = (await self._http.send_async(HttpRequest("GET", url))).ensure_success()
        root = resp.json()
        cur = root["current"]
        ts = cur.get("time")
        return WeatherSample(
            at_utc=parse_utc(ts),
            temp_c=float(cur["temperature_2m"]),
            feels_like_c=float(cur["apparent_temperature"]),
            precip_mm=float(cur["precipitation"]),
            wind_kph=float(cur["wind_speed_10m"]) * 3.6,  # m/s -> km/h
            cloud_pct=int(cur["cloud_cover"]),
            condition=_wmo_decode(int(cur["weather_code"])),
        )

    async def hourly_async(
        self, lat: float, lon: float, hours: int
    ) -> List[WeatherSample]:
        if hours <= 0 or hours > 168:
            raise ValueError("hours out of range (1..168)")
        url = (
            "https://api.open-meteo.com/v1/forecast"
            f"?latitude={_invariant(lat)}&longitude={_invariant(lon)}"
            "&hourly=temperature_2m,apparent_temperature,precipitation,"
            "wind_speed_10m,cloud_cover,weather_code"
            f"&forecast_hours={hours}"
        )
        resp = (await self._http.send_async(HttpRequest("GET", url))).ensure_success()
        root = resp.json()
        h = root["hourly"]
        time = h["time"]
        temp = h["temperature_2m"]
        feel = h["apparent_temperature"]
        prec = h["precipitation"]
        wind = h["wind_speed_10m"]
        cld = h["cloud_cover"]
        code = h["weather_code"]
        n = min(len(time), hours)
        result: List[WeatherSample] = []
        for i in range(n):
            result.append(
                WeatherSample(
                    at_utc=parse_utc(time[i]),
                    temp_c=float(temp[i]),
                    feels_like_c=float(feel[i]),
                    precip_mm=float(prec[i]),
                    wind_kph=float(wind[i]) * 3.6,
                    cloud_pct=int(cld[i]),
                    condition=_wmo_decode(int(code[i])),
                )
            )
        return result
